From 578b8d7dc5e3fc00d308660fa60bb529a2e42bb3 Mon Sep 17 00:00:00 2001
From: Collin Funk <collin.funk1@gmail.com>
Date: Thu, 1 Jan 2026 20:24:29 -0800
Subject: [PATCH] fts: Improve performance on the Lustre file system.

Origin: upstream, https://github.com/coreutils/gnulib/commit/578b8d7dc5e3fc00d308660fa60bb529a2e42bb3.patch
Bug: https://bugs.gnu.org/80106
Bug-Ubuntu: https://launchpad.net/gnulib/+bug/2137373
Bug-Debian: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1129092

Reported by Tim Day <timday@amazon.com> in
<https://bugs.gnu.org/80106>.

* lib/fts.c (S_MAGIC_LUSTRE): New macro.
(dirent_inode_sort_may_be_useful): Return false on Lustre file systems
so inode sorting is not performed.
---
 ChangeLog | 9 +++++++++
 lib/fts.c | 6 ++++++
 2 files changed, 15 insertions(+)

Index: coreutils-9.7/lib/fts.c
===================================================================
--- coreutils-9.7.orig/lib/fts.c
+++ coreutils-9.7/lib/fts.c
@@ -658,6 +658,7 @@ enum leaf_optimization
 /* Linux-specific constants from coreutils' src/fs.h */
 # define S_MAGIC_AFS 0x5346414F
 # define S_MAGIC_CIFS 0xFF534D42
+# define S_MAGIC_LUSTRE 0x0BD00BD0
 # define S_MAGIC_NFS 0x6969
 # define S_MAGIC_PROC 0x9FA0
 # define S_MAGIC_TMPFS 0x1021994
@@ -764,6 +765,11 @@ dirent_inode_sort_may_be_useful (FTSENT

   switch (filesystem_type (p, dir_fd))
     {
+    case S_MAGIC_LUSTRE:
+      /* On Lustre, sorting directory entries interferes with its ability to
+         prefetch file metadata (via statahead).  This would make a make a
+         command like 'du' around 9 times slower.  See
+         <https://bugs.gnu.org/80106>.  */
     case S_MAGIC_CIFS:
     case S_MAGIC_NFS:
     case S_MAGIC_TMPFS:
